src/matching.rs

```diff
@@ -144,14 +145,18 @@ impl<T, Impl: SelectorImpl> SelectorMap<T, Impl> {
         sort_by(&mut matching_rules_list[init_len..], &compare);
     }
 
-    fn get_matching_rules_from_hash<E, V>(element: &E,
-                                          parent_bf: Option<&BloomFilter>,
-                                          hash: &HashMap<Atom, Vec<Rule<T, Impl>>>,
-                                          key: &Atom,
-                                          matching_rules: &mut V,
-                                          shareable: &mut bool)
-                                          where E: Element<Impl=Impl>,
-                                                V: VecLike<DeclarationBlock<T>> {
+    fn get_matching_rules_from_hash<E, S, B: ?Sized, V>(
+        element: &E,
+        parent_bf: Option<&BloomFilter>,
+        hash: &HashMap<S, Vec<Rule<T, Impl>>>,
+        key: &B,
```

Comment:
Hmm... A bit more descriptive names for the types here wouldn't hurt, though feel free not to change them, or doing that in a followup.